ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/AbstractTaskFragment.kt

```diff
@@ -0,0 +1,148 @@
+    savedInstanceState: Bundle?
+  ): View? {
+    super.onCreateView(inflater, container, savedInstanceState)
+
+    @Suppress("UNCHECKED_CAST")
+    viewModel = dataCollectionViewModel.getTaskViewModel(position) as T
+
+    taskView = onCreateTaskView(inflater, container)
+    taskView.bind(this, viewModel)
+    taskView.addTaskView(onCreateTaskBody(inflater))
+
+    onCreateActionButtons()
+    onActionButtonsCreated()
+
+    return taskView.root
+  }
+
+  /** Creates the view for common task template with/without header. */
+  abstract fun onCreateTaskView(inflater: LayoutInflater, container: ViewGroup?): TaskView
+
+  /** Creates the view for body of the task. */
+  abstract fun onCreateTaskBody(inflater: LayoutInflater): View
+
+  /** Invoked when the fragment is ready to add buttons to the current [TaskView]. */
+  open fun onCreateActionButtons() {
+    addContinueButton()
+    addSkipButton()
+  }
+
+  /** Invoked when the all [ButtonAction]s are added to the current [TaskView]. */
+  open fun onActionButtonsCreated() {
+    viewModel.taskData.observe(viewLifecycleOwner) { onTaskDataUpdated(it.orElse(null)) }
+  }
+
+  /** Invoked when the data associated with the current task gets modified. */
+  protected open fun onTaskDataUpdated(taskData: TaskData?) {
+    for ((_, button) in buttons) {
+      button.onTaskDataUpdated(taskData)
+    }
+  }
+
+  private fun addContinueButton() {
+    addButton(ButtonAction.CONTINUE)
+      .setOnClickListener { dataCollectionViewModel.onContinueClicked() }
+      .setOnTaskUpdated { button, taskData ->
+        button.updateState { isEnabled = taskData.isNotEmpty() }
+      }
+      .updateState { isEnabled = false }
+  }
+
+  private fun addSkipButton() {
+    addButton(ButtonAction.SKIP)
+      .setOnClickListener {
+        viewModel.clearResponse()
+        dataCollectionViewModel.onContinueClicked()
+      }
+      .updateState { visibility = if (viewModel.isTaskOptional()) View.VISIBLE else View.GONE }
+  }
+
+  fun addUndoButton() {
+    addButton(ButtonAction.UNDO)
+      .setOnClickListener { viewModel.clearResponse() }
+      .setOnTaskUpdated { button, taskData ->
+        button.updateState { visibility = if (taskData.isEmpty()) View.GONE else View.VISIBLE }
+      }
+      .updateState {
+        visibility = View.GONE
+        isEnabled = true
+      }
+  }
+
+  private fun addButton(action: ButtonAction): TaskButton {
+    check(!buttons.contains(action)) { "Button $action already bound" }
+    val button =
+      TaskButton.createAndAttachButton(action, taskView.actionButtonsContainer, layoutInflater)
+    buttons[action] = button
+    return button
+  }
+
+  protected fun getButton(action: ButtonAction): TaskButton {
+    check(buttons.contains(action)) { "Expected key $action in $buttons" }
+    return buttons[action]!!
+  }
+}
+
+private fun TaskData?.isEmpty(): Boolean = this?.isEmpty() ?: true
+
+private fun TaskData?.isNotEmpty(): Boolean = !(this?.isEmpty() ?: true)
```

Comment:
Couldn't this simply be defined as `!this.isEmpty()`? Using the previous extension function?

Reply:
Done